### PR TITLE
Add Bitbucket and Bitbucket Server support

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
Add Bitbucket and Bitbucket Server support

### What issues does this PR fix or reference?

https://issues.redhat.com/browse/CRW-4323

### How to test this PR?

1. Deploy bitbucket server: https://github.com/che-incubator/git-srv/tree/main/bitbucket.
2. Create a repository with a devfile in the bitbucket server instance.
3. Go to projects directory: '/tools/devworkspace-generator' directory
4. execute `yarn`
5. execute 
```bash
node lib/entrypoint.js \
        --plugin-registry-url:https://che-plugin-registry-main.surge.sh/v3/ \
        --editor-entry:che-incubator/che-code/latest \
        --output-file:<output directory> \
        --injectDefaultComponent:true \
        --defaultComponentImage:registry.access.redhat.com/ubi8/openjdk-11:latest \
        --devfile-url:<url of the bitbucket server repo>
```
See: `yaml` file is generated and contains the devfile from the repository.
### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: tools/devworkspace-generator/src/main.ts [hunk 1]
```diff
@@ -69,12 +70,14 @@ export class Main {
     let devfileContent;
     let editorContent;
 
-    // gets the github URL
+    // gets the repo URL
     if (params.devfileUrl) {
-      const githubResolver = container.get(GithubResolver);
-      const githubUrl = githubResolver.resolve(params.devfileUrl);
+      // const githubResolver = container.get(GithubResolver);
+      // const githubUrl = githubResolver.resolve(params.devfileUrl);
```
Comment (svor) [inline tools/devworkspace-generator/src/main.ts:76]: let's remove them?

File: tools/devworkspace-generator/src/resolve/url.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/**********************************************************************
+ * Copyright (c) 2022 Red Hat, Inc.
```
Comment (svor) [inline tools/devworkspace-generator/src/resolve/url.ts:2]:
```suggestion
 * Copyright (c) 2023 Red Hat, Inc.
```

File: tools/devworkspace-generator/tests/bitbucket-server/bitbucket-server-resolver.spec.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
+/**********************************************************************
+ * Copyright (c) 2022 Red Hat, Inc.
```
Comment (svor) [inline tools/devworkspace-generator/tests/bitbucket-server/bitbucket-server-resolver.spec.ts:2]:
```suggestion
 * Copyright (c) 2023 Red Hat, Inc.
```

File: tools/devworkspace-generator/tests/bitbucket/bitbucket-resolver.spec.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
+/**********************************************************************
+ * Copyright (c) 2022 Red Hat, Inc.
```
Comment (svor) [inline tools/devworkspace-generator/tests/bitbucket/bitbucket-resolver.spec.ts:2]:
```suggestion
 * Copyright (c) 2023 Red Hat, Inc.
```

**Review by svor — APPROVED** [review]
In general codebase looks good to me

File: tools/devworkspace-generator/src/main.ts (hunk 1 above)
Comment (vinokurig) [reply tools/devworkspace-generator/src/main.ts]: removed the leftovers

File: tools/devworkspace-generator/src/resolve/url.ts (hunk 2 above)
Comment (vinokurig) [reply tools/devworkspace-generator/src/resolve/url.ts]: done

**Comment by vinokurig** [comment]
@svor I have added some new fixes for the `build/scripts/generate_devworkspace_templates.sh` and `build/scripts/update_devworkspace_templates.sh` scripts. Could you please take a look.

File: build/scripts/generate_devworkspace_templates.sh [hunk 5]
```diff
@@ -18,8 +18,11 @@ do
   if [ -n "$devfile_url" ]; then
     devfile_url=${devfile_url##*v2: }
     devfile_url=${devfile_url%/}
-    devfile_repo=${devfile_url%/tree*}
-    name=$(basename "${devfile_repo}")
+    npm_config_yes=true npx @eclipse-che/che-devworkspace-generator@${CHE_DEVWORKSPACE_GENERATOR_VERSION} \
```
Comment (svor) [inline build/scripts/generate_devworkspace_templates.sh:21]: could you please add a comment why do we need to generate temp.yaml file for each sample

File: build/scripts/generate_devworkspace_templates.sh [hunk 6]
```diff
@@ -18,8 +18,11 @@ do
...
+        --devfile-url:"${devfile_url}" \
+        --editor-entry:che-incubator/che-code/latest \
+        --output-file:"${dir}"temp.yaml
+    name=$(yq -r '.spec.template.projects[0].name' "${dir}temp.yaml"  | sed -n '2 p')
```
Comment (svor) [inline build/scripts/generate_devworkspace_templates.sh:25]: probably need to remove temp.yaml after getting name and url from it

**Review by svor — CHANGES_REQUESTED** [review]
I've tried to run ./build.sh script and it was failed because of test didn't cover new code, could you please fix that.

The GH didn't affect that because it builds only Dockerfile, probably we have to fix that but not in this PR

File: build/scripts/generate_devworkspace_templates.sh (hunk 5 above)
Comment (vinokurig) [reply build/scripts/generate_devworkspace_templates.sh:22]: done

File: build/scripts/generate_devworkspace_templates.sh (hunk 6 above)
Comment (vinokurig) [reply build/scripts/generate_devworkspace_templates.sh:26]: done

**Comment by vinokurig** [comment]
@svor I've fixed the tests so the `./build.sh` script executes without errors. Addressed other comments, please take a look.

**Review by svor — APPROVED** [review]
-  Tested for both Bitbucket and Github repositories.
-  Generated devworkspaces by:
```
node lib/entrypoint.js \
        --devfile-url:https://bitbucket-bitbucket.apps.cluster-84ndn.84ndn.sandbox2846.opentlc.com/scm/myp/ansible-devspaces-demo.git \
        --plugin-registry-url:https://che-plugin-registry-main.surge.sh/v3/ \
        --editor-entry:che-incubator/che-code/latest \
        --output-file:/tmp/devworkspace-che-code-ansible-latest.yaml
DevWorkspace che-code-ansible-demo was generated.
```
-  Started workspaces from generated files 
screenshot-eclipse-che apps cluster-84ndn 84ndn sandbox2846 opentlc com-2023 06 30-15_08_13

**Review by ScrewTSW — APPROVED** [review]
LGTM

**Review by ibuziuk — APPROVED** [review]

**Pull request merged**